src/providers/wfs/oapif/qgsoapifprovider.cpp

```diff
@@ -214,28 +214,54 @@ QgsWkbTypes::Type QgsOapifProvider::wkbType() const
 
 long long QgsOapifProvider::featureCount() const
 {
-  if ( mUpdateFeatureCountAtNextFeatureCountRequest )
+  // If no filter is set use the fast way of retrieving the feature count
+  if ( mSubsetString.isEmpty() )
   {
-    mUpdateFeatureCountAtNextFeatureCountRequest = false;
+    QString url = mShared->mItemsUrl;
+    url += QStringLiteral( "?limit=1" );
+    url = mShared->appendExtraQueryParameters( url );
+
 
-    QgsFeature f;
-    QgsFeatureRequest request;
-    request.setNoAttributes();
-    auto iter = getFeatures( request );
-    long long count = 0;
-    bool countExact = true;
-    while ( iter.nextFeature( f ) )
+    if ( !mShared->mServerFilter.isEmpty() )
     {
-      if ( count == 1000 ) // to avoid too long processing time
+      url += QStringLiteral( "&" );
+      url += mShared->mServerFilter;
+    }
+
+    QgsOapifItemsRequest itemsRequest( mShared->mURI.uri(), url );
+    if ( !itemsRequest.request( true, false ) )
+      return -1;
+    if ( itemsRequest.errorCode() != QgsBaseNetworkRequest::NoError )
+      return -1;
+
+    mShared->setFeatureCount( itemsRequest.numberMatched(), true );
```

Comment:
you cannot assume that the response will include a numberMatched property. Cf Requirement 31.B of https://docs.opengeospatial.org/is/17-069r4/17-069r4.html : "A server MAY omit this information in a response, if the information about the number of matching features is not known or difficult to compute.".
So if it is missing, the slow way should probably be done.

Reply:
@rouault thanks for the review. I changed the code so that at first (if no filter set) it tries to get the numberMatched parameter. If not available the old way is used

Reply:
> I changed the code so that at first (if no filter set) it tries to get the numberMatched parameter. If not available the old way is used

thanks. did you check that in the tests there's a scenario that tests this fallback branch?

Reply:
I have added a test now and discovered that the feature count was not updated until a first call to `reloadProviderData`. To fix that i changed the default value of this variable : https://github.com/qgis/QGIS/pull/51590/files#diff-1510a5eb3effae0928488d1df87da7abc5ad6ae1062a40f5777381b81db3f0f1R113